packages/appium/docs/en/ecosystem/build-drivers.md

```diff
@@ -2,4 +2,617 @@
+up in Appium, start and stop sessions with it, etc...
+
+To make it available to users, you could publish it via NPM. When you do so, your driver will be
+isntallable via the Appium CLI:
+
+```
+appium driver install --source=npm <driver-package-on-npm>
+```
+
+### Developing your driver
+
+How you develop your driver is up to you. It is convenient, however, to run it from within Appium
+without having to do lots of publishing and installing. The most straightforward way to do this is
+to include the most recent version of Appium as a `devDependency`, and then also your own driver,
+like this:
+
+```json
+{
+    "devDependencies": {
+        ...,
+        "appium": "^2.0.0",
+        "your-driver": "file:.",
+        ...
+    }
+}
+```
+
+Now, you can run Appium locally (`npx appium`), and because your driver is listed as a dependency
+alongside it, it will be automatically "installed" and available. You can design your e2e tests
+this way, or if you're writing them in Node.js, you can simply import Appium's start server methods
+to handle starting and stopping the Appium server in Node.
+
+## Standard driver implementation ideas
+
+These are things you will probably find yourself wanting to do when creating a driver.
+
+### Set up state in a constructor
+
+If you define your own constructor, you'll need to call `super` to make sure all the standard state
+is set up correctly:
+
+```js
+constructor(opts) {
+    super(opts);
+    // now do your own thing
+}
+```
+
+### Define and validate accepted capabilities
+
+You can define your own capabilities and basic validation for them. Users will always be able to
+send in capabilities that you don't define, but if they send in capabilities you have explicitly
+defined, then Appium will validate that they are of the correct type (and will check for the
+presence of required capabilities).
+
+If you want to turn capability validation off entirely, set `this.shouldValidateCaps` to `false` in
+your constructor.
+
+To give Appium your validation constraints, set `this.desiredCapConstraints` to a validation object
+in your constructor. Validation objects can be somewhat complex. Here's an example from the
+UiAutomator2 driver:
+
+```js
+{
+  app: {
+    presence: true,
+    isString: true
+  },
+  automationName: {
+    isString: true
+  },
+  browserName: {
+    isString: true
+  },
+  launchTimeout: {
+    isNumber: true
+  },
+}
+```
+
+### Start a session and read capabilities
+
+Appium's `BaseDriver` already implements the `createSession` command, so you don't have to. However
+it's very common to need to perform your own startup actions (launching an app, running some
+platform code, or doing different things based on capabilities you have defined for your driver).
+So you'll probably end up overriding `createSession`. You can do so by defining the method in your
+driver:
+
+```js
+async createSession(jwpCaps, reqCaps, w3cCaps, otherDriverData) {
```

Comment:
this is now different and all the first 3 params are assumed to be be `w3cCaps`

Reply:
maybe not everything was cut off. Basically the code there just checks which of these three arguments is a valid w3c caps one (starting from the last one) and accepts it or throws an exception if none is found. My point was we probably don't want to do a breaking change to all drivers and also don't wan't to expose legacy jwp support.

Reply:
yeah. i think all the info is still there from the client, but now it doesn't matter which one comes in which position as long as a w3c object is in there. but it does mean driver authors have to be careful when they are calling `super`, for example, not to send in 3 copies of the jwp style one